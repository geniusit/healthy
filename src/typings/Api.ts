export interface Api {
    name: string,
    endpoint: string,
    options?: {
        message: string,
    }
}
