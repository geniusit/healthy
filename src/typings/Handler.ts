import { Api } from './Api';
import { Response } from './Response';

export type Handler = (api: Api, response: Response) => void;
