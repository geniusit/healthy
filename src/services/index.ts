import { Api, Response as ApiResponse } from '../typings/Api';
import { Handler } from '../typings/Handler';
 
/** Begins the chain of API calls */
export const begin = async (apis: Api[], onResponse: Handler, interval?: number) => {
 const currentInterval = interval ? interval : 30000;
 
 apis.forEach(async a => {
   await makeCall(a, onResponse);
 });
 
 /** Set the timeout and do it again */
 setTimeout(async () => await begin(apis, onResponse, currentInterval), currentInterval);
};
 
/** Make the actual fetch and return a Response object */
const makeCall = async (api: Api, onResponse: Handler) => {
 await fetch(api.endpoint)
   .then(apiResponse => handleResponse(api, apiResponse, onResponse))
   // tslint:disable-next-line
   .catch(error => handleResponse(api, error, onResponse));
};
 
/**
* Handles the Response
*/
const handleResponse = (api: Api, response: Response, onResponse: Handler) => {
 const apiResponse: ApiResponse = {
   code: response.status,
   message: response.statusText,
 };
 
 onResponse(api, apiResponse);
};
 
export default { begin, makeCall, handleResponse };