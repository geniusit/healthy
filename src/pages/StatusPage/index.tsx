import React, { Component } from 'react';
import { Api, ApiWithResponse, Response } from '../../typings/Api';
import { begin } from '../../services';
import { Container } from './StatusPage.elements';
import { PageStatus, StatusRow } from '../../components';

type StatusPageProps = {
  /** An array of API objects */
  apis: Array<Api>;
  /** The interval at which to call the APIs in milliseconds */
  interval?: number;
};

interface ApiWithError extends ApiWithResponse {
  hasError: boolean;
}

type StatusPageState = {
  apis: Array<ApiWithError>;
  hasError: boolean;
};

class StatusPage extends Component<StatusPageProps> {
  state: StatusPageState = {
    apis: [],
    hasError: false,
  };

  handleError = (api: Api, response: Response) => {
    const apis = [...this.state.apis];
    const index = apis.findIndex(stateApi => stateApi.api === api);
    apis[index].hasError = true;
    apis[index].response = response;

    this.setState(
      {
        apis,
        hasError: true
      }
    );
  };

  componentDidMount = async () => {
    const apis = this.props.apis.map(api => ({ api, response: null, hasError: false }));

    this.setState({
      apis: apis,
      hasError: false,
    });

    await begin(this.props.apis, this.handleError, this.props.interval, undefined);
  };

  render() {
    return (
      <Container>
        <PageStatus hasError={this.state.hasError} />
        {this.state.apis.map((api, i) => (
          <StatusRow
            key={api.api.name + api.api.endpoint}
            name={api.api.name}
            hasError={api.hasError}
            className={i === 0 ? 'first' : ''}
          />
        ))}
      </Container>
    );
  }
}

export default StatusPage;
