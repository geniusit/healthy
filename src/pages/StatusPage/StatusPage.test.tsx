import React from 'react';
import { shallow } from 'enzyme';
import StatusPage from './';
import { Api, Response } from '../../typings/Api';
jest.mock('../../services/');
import { begin } from '../../services/';


describe('src/pages/StatusPage', () => {
    const apis: Array<Api> = [
        {
            name: 'Api 1',
            endpoint: 'https://google.com',
        },
        {
            name: 'Api 2',
            endpoint: 'https://yahoo.com'
        }
    ];

    describe('render', () => {
        describe('default render', () => {
            it('should match the snapshot', () => {
                const component = shallow(<StatusPage apis={apis}/>);
                expect(component).toMatchSnapshot();
            })
        })
    });

    describe('handleError', () => {
        it('should update the api in state', async () => {
            const component = shallow<StatusPage>(<StatusPage apis={apis} />);
            const mockResponse: Response = {
                code: 500,
                message: 'It broke'
            }
            
            await component.instance().handleError(apis[0], mockResponse);

            expect(component.state('apis')).toEqual([
                {
                    api: {
                        name: 'Api 1',
                        endpoint: 'https://google.com',
                    },
                    response: {
                        code: 500,
                        message: 'It broke'
                    },
                    hasError: true
                },
                {
                    api: {
                    name: 'Api 2',
                    endpoint: 'https://yahoo.com',
                    },
                    response: null,
                    hasError: false
                }
            ])
        })
    })

    describe('componentDidMount', () => {
        it('should update state from props', async () => {
            const component = shallow<StatusPage>(<StatusPage apis={apis} />);
            await component.instance().componentDidMount();

            expect(component.state('apis')).toEqual([
                {
                    api: {
                        name: 'Api 1',
                        endpoint: 'https://google.com',
                    },
                    response: null,
                    hasError: false
                },
                {
                    api: {
                    name: 'Api 2',
                    endpoint: 'https://yahoo.com',
                    },
                    response: null,
                    hasError: false
                }
            ])
        })

        it('should call begin', async () => {
            const component = shallow<StatusPage>(<StatusPage apis={apis} />);
            await component.instance().componentDidMount();

            expect(begin).toHaveBeenCalled();
        })
    })
})