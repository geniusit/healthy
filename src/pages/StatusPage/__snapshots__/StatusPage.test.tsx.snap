// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/pages/StatusPage render default render should match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <StatusPage
    apis={
      Array [
        Object {
          "endpoint": "https://google.com",
          "name": "Api 1",
        },
        Object {
          "endpoint": "https://yahoo.com",
          "name": "Api 2",
        },
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <PageStatus
          hasError={false}
        />,
        Array [
          <StatusRow
            className="first"
            hasError={false}
            name="Api 1"
          />,
          <StatusRow
            className=""
            hasError={false}
            name="Api 2"
          />,
        ],
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "hasError": false,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "Api 1https://google.com",
        "nodeType": "function",
        "props": Object {
          "className": "first",
          "hasError": false,
          "name": "Api 1",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "Api 2https://yahoo.com",
        "nodeType": "function",
        "props": Object {
          "className": "",
          "hasError": false,
          "name": "Api 2",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": Object {
      "$$typeof": Symbol(react.forward_ref),
      "__emotion_base": "div",
      "__emotion_forwardProp": undefined,
      "__emotion_real": [Circular],
      "__emotion_styles": Array [
        "
  width: 100%;
  max-width: 800px;
  margin: 60px auto;
",
      ],
      "defaultProps": undefined,
      "displayName": "Styled(div)",
      "render": [Function],
      "withComponent": [Function],
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <PageStatus
            hasError={false}
          />,
          Array [
            <StatusRow
              className="first"
              hasError={false}
              name="Api 1"
            />,
            <StatusRow
              className=""
              hasError={false}
              name="Api 2"
            />,
          ],
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "hasError": false,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "Api 1https://google.com",
          "nodeType": "function",
          "props": Object {
            "className": "first",
            "hasError": false,
            "name": "Api 1",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "Api 2https://yahoo.com",
          "nodeType": "function",
          "props": Object {
            "className": "",
            "hasError": false,
            "name": "Api 2",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": Object {
        "$$typeof": Symbol(react.forward_ref),
        "__emotion_base": "div",
        "__emotion_forwardProp": undefined,
        "__emotion_real": [Circular],
        "__emotion_styles": Array [
          "
  width: 100%;
  max-width: 800px;
  margin: 60px auto;
",
        ],
        "defaultProps": undefined,
        "displayName": "Styled(div)",
        "render": [Function],
        "withComponent": [Function],
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
